### PR TITLE
Allow events to propagate up to the ListView to regain onItemClick and onItemLongClick

### PR DESCRIPTION
**Pull request by mcumings**
This approach is counter-intuitive but I locally verified that the ListView click handlers
work while also verifying that the Button on the list items can receive their click events
as well.

### CODE REVIEW

**Comment by wdullaer** [comment]
This is a really tricky one.
If you do not have a clickable item in your listrow, everything works as expected.
If you do, it seems that onListItemClicked not longer fires, even if you don't click the button.

Your fix does work in touch mode: onListItemClicked fires and the button can be clicked, but stopping the descendants from obtaining focus probably means that you can no longer select the button on the row with a keyboard. I have a feeling that this might have other accessibility implications as well.

The solution is good, but not ideal. It should be possible to fix this another way, but it escapes me right now.

**Comment by mcumings** [comment]
Yeah, I tend to agree.  It's one of those things you code that you're not 100% self-confident in. :)  I don't have an accessibility test setup readily available at the moment to try it out.  I did try keyboard-based navigation (via emulation) and it worked as well as the example app did previously (read: I wasn't able to figure out how to focus the Button) but I'd want to try gesture nav or TalkBack.

It's looking like we may not end up using this library after all.  Your call as to whether you'd like to keep this PR open or not.

Thanks for considering it.

**Comment by wdullaer** [comment]
I'll probably merge this in later today. It is a lot better than the current situation.

I think I know which bit of code is to blame, but I'm not sure how I can fix it properly at this point in time.

Also: can I ask why you're no longer considering the library?

**Comment by mcumings** [comment]
The main reason that I believe we're moving away from the library has less to do with the library and more to do with changing requirements.  That being said here are the areas that I encountered that were difficult for our use cases as they were originally formulated:

* The revealed views are static and there weren't any easy hooks for manipulating those views.  We have use cases where the swipe behavior acts as a toggle (e.g., "read" to/from "unread") and we need to be able to change the text, color, etc. from list item to list item.
* I didn't spend a lot of time digging but it seemed like it would be difficult to implement/customize our own animations and translations.  We wanted a soft-swipe to reveal two buttons for performing actions like the RecyclerView demo here: https://github.com/daimajia/AndroidSwipeLayout.
* There was a potential that we wanted to swipe from either side to perform different actions where the actions on either side would be enabled/disabled based on the list item data.  If we did that then I wanted to add a translation that expressed resistance when the action was unsupported, such as when attempting to left-swipe to set a reminder on a collection in Google's Inbox app.

Rather than only list the negatives, here are some positives:
* For what the library was designed to do, it did it very well.
* Support for both swiping and far swiping, with easy dismissal, etc.
* Compared to most other swipe libraries, the integration by wrapping the adapter was hugely welcomed.
* The code was simpler and cleaner than the other libraries I had evaluated and it was easy to understand what was being expressed.

Who knows.. The requirements may change again and I'll be back for more. :)  Thanks for publishing a solid library!

**Comment by wdullaer** [comment]
I appreciate the feedback. I'm glad you found it well designed.
I was really just scratching my own itch here, and thought someone else might find it useful.

Regarding your changed requirements:
* **Static backgrounds**
The backgrounds are indeed static. On top of my head, I can't think of an elegant way to make them easy to change, especially because the logic which will determine which background to show is most likely highly different for each app. If you want this it is probably best to extend the class and override the `getView(final int position, final View convertView, final ViewGroup parent)`method so that it adds the backgrounds you need to the items based on your logic

* **Stopping swipes**
This library is indeed not really meant to show buttons on a swipe. People seem to have had success with https://github.com/baoyongzhang/SwipeMenuListView, but I haven't used it myself.
RecyclerView exists because ListViews are quite hard to animate. Once I figure out how to make them play nice with Cursors, I plan to migrate my apps to them.

* **Disable swipe based on direction**
You are not the first one to ask for that feature. If I want to do that I'll need to change the place where `hasActions`is called. At the current point in time I do not know the swipe direction yet. However, when I know the swipe direction, the row item will have already moved a bit, or the swipe will feel laggy. Having it resist a swipe, like in Inbox, is a good way to mask this though.
In short: I think this one can be incorporated (I'm mostly scared about the "resistance" animation, I'm not particularly good at coding those)
